### DISCUSSION TITLE
Decoupling billboards from render material

### DISCUSSION THREAD

**Discussion by QbieShay** [godotengine/godot-proposals#6605] [3D]
I find it incredibly clunky that billboard modes need to be reimplemented per-shader. I believe that it should be something separate, that's either done automagically at a particle stage and not at the render stage, or that can be just selected on a material, even on shaders. it's very annoying to work with right now.

**Comment by Calinou** [comment]
If you want billboarding to still be performed on the GPU (which is a requirement for GPU-based particles), I don't see a way to do this until the vertex and fragment shader stages can be loaded separately. Note that even if this is done, there won't be a way to use other vertex shader effects if you're overriding the billboard mode.

**Reply by QbieShay** [reply]
I disagree. I think it should be possible to glue code after the user vertex stage to make it work, like Godot adds its own code to do skinning etc.

**Reply by QbieShay** [reply]
Or alternatively as a last step before passing the transforms to the render shader

**Reply by AThousandShips** [reply]
It would have to be applied to the matrices *before* the vertex shader no?

**Comment by AThousandShips** [comment]
I don't see how making this less flexible is helpful, then if you want more specific modifications to this you have to write your own billboard code without the aid of the one provided from material

Unlike other features like skinning this is an area where flexibility is relevant, and this won't really reduce any complexity of the shader code or the compilation

**Reply by QbieShay** [reply]
You'd still be able to make your own if you wish to do so - just by disabling the built-in billboard. I don't think this is less flexible

**Reply by AThousandShips** [reply]
It does make it harder for users to make modifications to the billboard code, as users will have to find it out on their own and add it instead of having it provided by a mateeial, now this can be helped by adding some example code to the documentation

But the only real difference from moving it is the minor effect of the compilation of the gdshader code, and I don't think that really makes any significant difference, and makes shaded variation more complicated

I don't really see how it's in the way to have just a snippet in the shader you can just ignore

**Reply by QbieShay** [reply]
> It does make it harder for users to make modifications to the billboard code, as users will have to find it out on their own and add it instead of having it provided by a mateeial

That's what i am arguing, having it done by the material and not by the user.

>But the only real difference from moving it is the minor effect of the compilation of the gdshader code,

No that's the opposite. It's to not have to constantly complicate any shader i want to use with billboard modes. It's not about compilation speed.

> I don't really see how it's in the way to have just a snippet in the shader you can just ignore
It's a lot of wasted time and clutter in the shader

**Reply by QbieShay** [reply]
Furthermore, did you implement a lot of custom billboard modes (something new, not copied from the base shaders)? I have and this would be 100% compatible with it.

**Reply by AThousandShips** [reply]
Well that's good, my bad

**Reply by QbieShay** [reply]
For context I work for a VFX artist and I wrote this post at the 4th time I have to pollute some other shader with billboard modes so that I can use it with particles. I have to go and modify shaders that are not written by me and that are unrelated to particles, just to add billboard code. It's repetitive, it clutters other people's code, and I don't see a good reason for doing it over and over.
